### PR TITLE
chore(db/migration): check migration regex path before db changes

### PR DESCRIPTION
**Pull request by chronolaw**
### Checklist

- [ ] The Pull Request has tests
- [x] There's an entry in the CHANGELOG
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

In `kong migrations up`, it will print an error log to notify the user if there is an incompatible regex path route:

```
2022/12/30 09:27:37 [error] Regex path may not work with router flavor 'traditional_compatible', route id: 0950d523-aa5e-4d05-a1af-928ea5c3aa99, err:  --> 1:2
  |
1 | (http.path ~ "^/\\/*/user$")
  |  ^------------------------^
  |
  = regex parse error:
    ^/\/*/user$
      ^^
error: unrecognized escape sequence
```

It will replace #9992 #10089.

Reference: https://github.com/Kong/kong/commit/f66c3ecb1bd8e335e20b6f751f71ef3b55085345

See KAG-329, we also need a user document to tell user how to solve these errors.

### CODE REVIEW

File: kong/db/migrations/validate_path_280_300.lua [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+
+local validate_atc_expression
+do
+  local router = require("resty.router.router")
+  local CACHED_SCHEMA = require("kong.router.atc").schema
+  local get_expression = require("kong.router.compat").get_expression
+
+  validate_atc_expression = function(route)
+    local r = router.new(CACHED_SCHEMA)
```
Comment (dndx) [inline kong/db/migrations/validate_path_280_300.lua:18]: We can cache the `router`.

```
if not r then
  r = router.new(CACHED_SCHEMA)
end

...

local res, err = r:add_matcher(0, route.id, exp)
if not res then
   log.error("Regex path will not work with the default router flavor in 3.0, " ..
                 "route id: %s, err: %s", route.id, err)
   return false
end

r:remove_matcher(route.id)
```

File: kong/db/migrations/validate_path_280_300.lua [hunk 2]
```diff
@@ -0,0 +1,121 @@
+local log = require "kong.cmd.utils.log"
+
+local assert = assert
+local ipairs = ipairs
+local migrate_path = require "kong.db.migrations.migrate_path_280_300"
+
+
+local MAX_COUNT = 10
```
Comment (dndx) [inline kong/db/migrations/validate_path_280_300.lua:8]:
```suggestion
local MAX_ERRORS_AT_A_TIME = 10
```

File: kong/db/migrations/validate_path_280_300.lua (hunk 1 above)
Comment (chronolaw) [reply kong/db/migrations/validate_path_280_300.lua]: done

File: kong/db/migrations/validate_path_280_300.lua (hunk 2 above)
Comment (chronolaw) [reply kong/db/migrations/validate_path_280_300.lua]: done

File: kong/db/migrations/validate_path_280_300.lua (hunk 1 above)
Comment (chronolaw) [reply kong/db/migrations/validate_path_280_300.lua]: This improvement can be port to master branch later.

**Review by dndx — APPROVED** [review]

**Pull request merged**